File: src/ios/node.m
    /********* node.m Cordova Plugin Implementation *******/
// cordova platform add ios
#import "node.h"
#import "FavorX.framework/Headers/FavorX.h"

@implementation node

static MobileNode *mobileNode = nil;

- (void)getVersion:(CDVInvokedUrlCommand*)command
{
    NSString *version = MobileVersion();
    CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:version];
    [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
}
- (void)start:(CDVInvokedUrlCommand*)command
{
    // Check command.arguments here.
    [self.commandDelegate runInBackground:^{
        NSDictionary* paramsDic = [command.arguments objectAtIndex:0];
        if (paramsDic.count > 0) {

            MobileOptions* options = [[MobileOptions alloc] init];
            // api setting
            options.apiPort = [[paramsDic objectForKey:@"APIPort"] intValue];
            options.debugAPIPort = [[paramsDic objectForKey:@"DebugAPIPort"] intValue];
            options.websocketPort = [[paramsDic objectForKey:@"WebsocketPort"] intValue];
            options.enableDebugAPI = [[paramsDic objectForKey:@"EnableDebugAPI"] boolValue];
            // p2p setup
            options.networkID = [[paramsDic objectForKey:@"NetworkID"] intValue];
            options.p2PPort = [[paramsDic objectForKey:@"P2PPort"] intValue];
            options.welcomeMessage = [paramsDic objectForKey:@"WelcomeMessage"];
            // kademlia
            options.binMaxPeers = [[paramsDic objectForKey:@"BinMaxPeers"] intValue];
            options.lightMaxPeers = [[paramsDic objectForKey:@"LightMaxPeers"] intValue];
            // cache size
            options.cacheCapacity = [[paramsDic objectForKey:@"CacheCapacity"] intValue];
            //node bootstrap
            options.bootNodes = [paramsDic objectForKey:@"BootNodes"];
            options.enableDevNode = [[paramsDic objectForKey:@"EnableDevNode"] boolValue];
            options.enableFullNode = [[paramsDic objectForKey:@"EnableFullNode"] boolValue];
            // chain setting
            options.chainEndpoint = [paramsDic objectForKey:@"ChainEndpoint"];
            options.oracleContract = [paramsDic objectForKey:@"OracleContract"];
            // domain resolver
            options.resolverOptions = [paramsDic objectForKey:@"ResolverOptions"];
            // security
            options.password = [paramsDic objectForKey:@"Password"];
            options.dataPath = [paramsDic objectForKey:@"DataPath"];
            // leveldb opts
            options.blockCacheCapacity = [[paramsDic objectForKey:@"BlockCacheCapacity"] intValue];
            options.openFilesLimit = [[paramsDic objectForKey:@"OpenFilesLimit"] intValue];
            options.writeBufferSize = [[paramsDic objectForKey:@"WriteBufferSize"] intValue];
            options.disableSeeksCompaction = [[paramsDic objectForKey:@"DisableSeeksCompaction"] boolValue];
            // misc
            options.verbosity = [paramsDic objectForKey:@"Verbosity"];

            // 获取Documents目录
            //NSString *docPath = [NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES) lastObject];
            //options.dataPath = docPath;
            NSError* error = nil;
            mobileNode = MobileNewNode(options, &error);
            if(error == nil){
                 // Some blocking logic...
                 CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:[paramsDic objectForKey:@"APIPort"]];
                // The sendPluginResult method is thread-safe.
                 [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
              }else{
                 CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR];
                 [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
              }
        }else {
            CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR];
            [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
        }

    }];
}
@end

File: www/node.js
var exec = require('cordova/exec');


exports.getVersion = function (success, error) {
    exec(success, error, "node", 'getVersion');
}

exports.start = function (success, error) {

    exec(success, error, "node", 'start', [{
        APIPort: 1633,
        DebugAPIPort: 1635,
        WebsocketPort: 1639,
        EnableDebugAPI: true,
        NetworkID: 19,
        P2PPort: 1634,
        WelcomeMessage: "",
        BinMaxPeers: 50,
        LightMaxPeers: 100,
        CacheCapacity: 4 * 1024,
        BootNodes:
            "/ip4/107.167.2.7/tcp/1809/p2p/12D3KooWEw5bXwg4ho63XSJCtmYdgXbVGUKrf1Uhpg3RbogLswxt",
        EnableDevNode: false,
        EnableFullNode: false,
        ChainEndpoint: "https://polygon-testnet.public.blastapi.io",
        OracleContract: "0x21aC8FE412Fd058eD29a67a69c81EF08fA34f443",
        ResolverOptions:
            `A959836a03abbc7bF7EFAD8F9422456150bFE567@https://polygon-testnet.public.blastapi.io`,
        Password: "123456",
        // eslint-disable-next-line no-undef
        DataPath: cordova.file.dataDirectory.substring(7),
        BlockCacheCapacity: 8 * 1024 * 1024,
        OpenFilesLimit: 200,
        WriteBufferSize: 4 * 1024 * 1024,
        DisableSeeksCompaction: false,
        Verbosity: "trace",
    },]);
}
